playground/backend/internal/cache/redis_cache.go

```diff
@@ -0,0 +1,148 @@
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	pb "beam.apache.org/playground/backend/internal/api"
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/go-redis/redis/v8"
+	"github.com/google/uuid"
+	"time"
+)
+
+const defaultExpirationTime = time.Minute * 15
+
+type RedisCache struct {
+	redisClient *redis.Client
+}
+
+// newRedisCache returns Redis implementation of Cache interface.
+// In case of problem with connection to Redis returns error.
+func newRedisCache(addr string) (*RedisCache, error) {
+	// ctx should be as a parameter
+	ctx := context.Background()
+
+	rc := RedisCache{redisClient: redis.NewClient(&redis.Options{Addr: addr})}
+	_, err := rc.redisClient.Ping(ctx).Result()
+	if err != nil {
+		return nil, err
+	}
+	return &rc, nil
+}
+
+func (rc *RedisCache) GetValue(pipelineId uuid.UUID, subKey SubKey) (interface{}, error) {
+	// ctx should be as a parameter
+	ctx := context.Background()
+
+	switch subKey {
+	case SubKey_Status:
+		var result pb.Status
+		err := get(ctx, rc.redisClient, pipelineId.String(), subKey, &result)
+		if err != nil {
+			return nil, err
+		}
+		return result, nil
+	default:
+		var result string
+		err := get(ctx, rc.redisClient, pipelineId.String(), subKey, &result)
+		if err != nil {
+			return nil, err
+		}
+		return result, nil
+	}
+}
+
+func (rc *RedisCache) SetValue(pipelineId uuid.UUID, subKey SubKey, value interface{}) error {
+	// ctx should be as a parameter
+	ctx := context.Background()
+
+	return set(ctx, rc.redisClient, pipelineId.String(), subKey, value)
+}
+
+func (rc *RedisCache) SetExpTime(pipelineId uuid.UUID, expTime time.Duration) error {
+	// ctx should be as a parameter
+	ctx := context.Background()
+
+	exists, err := rc.redisClient.Exists(ctx, pipelineId.String()).Result()
+	if err != nil {
+		return err
+	}
+	if exists == 0 {
+		return fmt.Errorf("key: %s doesn't exist", pipelineId)
+	}
+
+	rc.redisClient.Expire(ctx, pipelineId.String(), expTime)
+	return nil
+}
+
+// set puts value to Redis by key and subKey (key:subKey:value).
+// If it is the first value for this key then set expiration for key.
+func set(ctx context.Context, redisClient *redis.Client, key string, subKey SubKey, value interface{}) error {
+	subKeyMarsh, err := marshal(subKey)
+	if err != nil {
+		return err
+	}
+	valueMarsh, err := marshal(value)
+	if err != nil {
+		return err
+	}
+	_, err = redisClient.HSet(ctx, key, subKeyMarsh, valueMarsh).Result()
+	if err != nil {
+		return err
+	}
+
+	redisClient.Expire(ctx, key, defaultExpirationTime)
+	return nil
+}
+
+// get sets value by key and sybKey to result variable with converting data by subKey.
```

Comment:
looks very strange when the method get _sets_ something

Reply:
Changed.